### Linear Regression with Coefficient and RMSE

## Merge feature and test datasets

In [33]:
### Importing the required packages and libraries
# we will need numpy and pandas later
import numpy as np
import pandas as pd


from sklearn.linear_model import LinearRegression
#import DecisionTreeClassifier class
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
# Randomly splitting the original dataset into training set and testing set
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

In [34]:
# reading a CSV file directly from Web, and store it in a pandas DataFrame:
# "read_csv" is a pandas function to read csv files from web or local device:
# walmart_features_df =  pd.read_csv('features.csv')

# walmart_sampleSubmission_df =  pd.read_csv('sampleSubmission.csv')

# walmart_stores_df =  pd.read_csv('stores.csv')

# walmart_test_df =  pd.read_csv('test.csv')

# walmart_train_df =  pd.read_csv('train.csv')

In [35]:
def mergeData(df):
    features =pd.read_csv('features.csv')
    storesdata =pd.read_csv('stores.csv')
    df = pd.merge(df, features, on=['Store','Date','IsHoliday'],how='inner')
    df = pd.merge(df, storesdata, on=['Store'], how='inner')
    return df



In [36]:
merged_df = mergeData(pd.read_csv('train.csv'))

In [37]:
merged_df.fillna(value=0, inplace=True)

In [38]:
merged_df['Markdowns'] = merged_df['MarkDown1'] + merged_df['MarkDown2'] + merged_df['MarkDown3'] + merged_df['MarkDown4'] + merged_df['MarkDown5'] 
labelsToDrop = ['MarkDown1', 'MarkDown2', 'MarkDown3','MarkDown4','MarkDown5']
merged_df.drop(labels=labelsToDrop,axis=1, inplace=True)

In [39]:
print(df.head())

   Store  Dept  Weekly_Sales  IsHoliday  Temperature  Fuel_Price         CPI  \
0      1     1      24924.50      False        42.31       2.572  211.096358   
1      1     2      50605.27      False        42.31       2.572  211.096358   
2      1     3      13740.12      False        42.31       2.572  211.096358   
3      1     4      39954.04      False        42.31       2.572  211.096358   
4      1     5      32229.38      False        42.31       2.572  211.096358   

   Unemployment Type    Size  Markdowns  Year  Month  Day  
0         8.106    A  151315        0.0  2010      2    5  
1         8.106    A  151315        0.0  2010      2    5  
2         8.106    A  151315        0.0  2010      2    5  
3         8.106    A  151315        0.0  2010      2    5  
4         8.106    A  151315        0.0  2010      2    5  


In [40]:
from pandas import DatetimeIndex
df = merged_df
df.Date = pd.to_datetime(df.Date)

#### fix date into two columns one for month and other for year


In [41]:
df['Year'] = DatetimeIndex(df['Date']).year
df['Month']= DatetimeIndex(df['Date']).month
df['Day'] = DatetimeIndex(df['Date']).day
df = df.drop(columns=['Date'])
df[0::1000]

Store  Dept  Weekly_Sales  IsHoliday  Temperature  Fuel_Price  \
0           1     1      24924.50      False        42.31       2.572   
1000        1    81      31614.83      False        72.55       2.835   
2000        1    95     131288.83      False        87.00       2.692   
3000        1    81      29383.02      False        51.41       2.771   
4000        1    79      30624.72      False        62.90       3.065   
5000        1    72      34638.69      False        83.00       3.699   
6000        1    85       2681.87       True        76.00       3.546   
7000        1    60       1443.20      False        51.63       3.159   
8000        1    44       4356.74      False        65.93       3.787   
9000        1    56       2159.46      False        84.88       3.286   
10000       1    52       2373.49      False        68.55       3.617   
11000       2    32      11575.39      False        68.07       2.808   
12000       2    46      27785.25      False        87.66       2.608   
13000       2    54        137.92      False        66.24       2.716   
14000       2    42       6637.94      False        38.25       2.989   
15000       2    33       5311.27      False        74.61       3.899   
16000       2    44       6984.53      False        89.86       3.554   
17000       2    38      63560.48       True        56.36       3.236   
18000       2    24       5935.40      False        58.79       3.630   
19000       2    25      15163.48      False        78.69       3.452   
20000       2    22      11627.47      False        75.88       3.717   
21000       3     7       7016.69      False        65.56       2.719   
22000       3    95      14751.85      False        84.32       2.623   
23000       3    90       1161.05      False        62.94       2.689   
24000       3    56       2111.99      False        67.40       3.065   
25000       3    44       2580.86      False        86.96       3.637   
26000       3    74       3121.86      False        75.54       3.285   
27000       3    13      11062.34      False        58.06       3.290   
28000       3    90       1693.11      False        75.19       3.688   
29000       3    85        198.70      False        85.18       3.638   
...       ...   ...           ...        ...          ...         ...   
392000     42     7        949.75      False        90.32       3.084   
393000     42    79      14375.51      False        58.83       3.047   
394000     42    87        914.84      False        61.05       4.089   
395000     42    31         52.50       True        89.06       3.771   
396000     42    67       1098.22      False        53.59       3.675   
397000     42    13      14851.51      False        88.58       4.144   
398000     42    82       7825.99      False        70.50       4.301   
399000     43    91      39088.11      False        88.48       2.653   
400000     43    91      30713.10      False        54.96       2.771   
401000     43     2      21153.45      False        74.49       3.807   
402000     43    21       1824.01      False        77.17       3.526   
403000     43    44          3.97       True        46.54       3.409   
404000     43    97      20529.81      False        82.68       3.227   
405000     44    95      28426.77      False        40.65       2.681   
406000     44    16       1127.05      False        78.94       2.797   
407000     44    74        322.53      False        35.21       2.842   
408000     44    97       6362.47      False        47.75       3.744   
409000     44    91      18664.95      False        63.60       3.648   
410000     44    82       3693.90       True        33.73       3.116   
411000     44    28         54.80      False        82.00       3.681   
412000     45     1      22416.94       True        27.73       2.773   
413000     45    83       1276.92      False        62.62       2.958   
414000     45    91      15753.50      False        78.37       2.

### Split the merged data set into X_train and y_train, use test dataset for X_test

In [48]:
#split with method

y =df['Weekly_Sales']
X = df.drop(columns=['Weekly_Sales','Type'])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, 
                                                    random_state=2)

In [55]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(X_train.head())

(295099, 12)
(295099,)
(126471, 12)
        Store  Dept  IsHoliday  Temperature  Fuel_Price         CPI  \
154115     16    98      False        31.73       3.479  195.718471   
238177     25    81      False        57.17       2.958  203.877024   
407348     44    91      False        23.35       2.974  127.719581   
6325        1    28      False        71.74       3.274  217.355273   
132096     14    22      False        45.01       3.582  184.613419   

        Unemployment    Size  Markdowns  Year  Month  Day  
154115         6.232   57197    7502.89  2011     11   18  
238177         7.856  128107       0.00  2010      5   21  
407348         7.224   39910       0.00  2011      2    4  
6325           7.866  151315       0.00  2011     10   14  
132096         8.549  200898       0.00  2011      3   11  


In [50]:
#Linear Regression

linear_reg = LinearRegression()
linear_reg.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [51]:
print("Intercept: ", linear_reg.intercept_)

print("Coefficient: " , linear_reg.coef_)

Intercept:  1366427.6184769147
Coefficient:  [-8.73094172e+01  1.10841005e+02  7.05852062e+02  1.98428062e+01
  2.92467634e+02 -2.08583897e+01 -1.83231124e+02  8.58850479e-02
  3.79210892e-02 -6.77895285e+02  1.30091858e+02 -1.54005316e+01]


### predictive model

$$y = 1366427.618 - 87.309 \times Store + 101.841 \times Department + 705.852 \times Is Holiday+ 19.842 \times Temperature + 292.467 \times FuelPrice - 20.858 \times CPI - 183.231 \times Unemployment + 0.0859 \times Size + 0.0379 \times Markdowns - 677.895 \times Year + 130.009 \times Month - 15.400 \times Day $$

###  Most important feature is 'IsHoliday'
### meaning that whether that specific day is a holiday or not has the greatest impact on weekly sales
### Second most important is Fuel Price follwed by Unemployment
### least important is markdowns but a lot of the data was 0

### Testing and prediction

In [52]:
y_pred = linear_reg.predict(X_test)
print(y_pred)

[10912.66259985 22031.38094417 18226.6893131  ...  4949.20422364
 17206.18091831 18621.10256506]


### RMSE

In [53]:
from sklearn import metrics
mse = metrics.mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)

In [54]:
print(rmse)

21830.539350501567
